### PR TITLE
graphdriver: Fix RefCounter memory leak

### PR DESCRIPTION
**Pull request by wcwxyz**
this fixes an issue where count can become -1 when a new
daemon instance attempts to unmount the container base dir, with the
later not being actually mounted. The most common action leading to this
scenario is an unorderly restart of the server running docker.

After count becomes -1, a combination of "docker start" + "docker
restart" will lead to a situation in which the mountpoint is not visible
from the global namespace, and attempts to run new processes in the
container will fail with an error like this one:

```
error: code = 2 desc = oci runtime error: exec failed:
container_linux.go:247: starting container process caused
"process_linux.go:75: starting setns process caused "fork/exec/proc/self/exe: no such file or directory"
```

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/graphdriver/counter.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ func (c *RefCounter) incdec(path string, infoOp func(minfo *minfo)) int {
 	}
 	infoOp(m)
 	count := m.count
+	if count <= 0 {
```
Comment (cpuguy83) [inline daemon/graphdriver/counter.go]: I think this makes more sense in `Decrement`?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Minor nit

File: daemon/graphdriver/counter.go (hunk 1 above)
Comment (wcwxyz) [reply daemon/graphdriver/counter.go]: Thanks for review.

It's doable. But `infoOp` should only handles `minfo *`. I think Function `incdec` seems a general place to take care of cleanup. Unless we want a another refactor of the whole code.

I think this fix is pretty simple. How do you want me to handle that?

Comment (cpuguy83) [reply daemon/graphdriver/counter.go]: It's fine, was just a suggestion.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by wcwxyz** [comment]
I don't know what's wrong with `Jenkins build Docker-PRs-WoW-RS1`. I checked the log and the failure didn't seem to be related with this patch. Can anyone familiar with this take a look please? Thanks.

**Comment by thaJeztah** [comment]
Looks like that test is flaky again on Windows 😞 opened an issue for that https://github.com/moby/moby/issues/36350, I'll restart Windows CI

```
03:25:44 ----------------------------------------------------------------------
03:25:44 FAIL: docker_api_logs_test.go:152: DockerSuite.TestLogsAPIUntil
03:25:44 
03:25:44 docker_api_logs_test.go:175:
03:25:44     c.Assert(len(allLogs), checker.GreaterOrEqualThan, 3)
03:25:44 ... obtained int = 1
03:25:44 ... expected int = 3
```

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**